### ISSUE TITLE
With `(foo)+` the corresponding sub-captures aren't removed during backtracking

### ISSUE THREAD

**Issue by raiph** [rakudo/rakudo#4105]
## The Problem

Sub-captures aren't removed during backtracking.

See https://stackoverflow.com/questions/65226308/why-do-i-get-different-backtracking-with-these-raku-regexes

## Expected Behavior
```
'abc' ~~ m/(\w)+ {say $/} bc/;

｢abc｣
 0 => ｢a｣
 0 => ｢b｣
 0 => ｢c｣
｢ab｣
 0 => ｢a｣
 0 => ｢b｣
｢a｣
 0 => ｢a｣
```

## Actual Behavior
```
｢abc｣
 0 => ｢a｣
 0 => ｢b｣
 0 => ｢c｣
｢ab｣
 0 => ｢a｣
 0 => ｢b｣
 0 => ｢c｣
｢a｣
 0 => ｢a｣
 0 => ｢b｣
 0 => ｢c｣
```

## Comments

I imagine the engine implements backtracking over a capture quantified by `*`, `+`, or `**` with a range by doing something like decrementing the outer `Match` object's `.pos` by the length of the last capture in the list of captures. Why doesn't it also do the equivalent of `capture[*-1]:delete`?

**Comment by jubilatious1** [comment]
Certainly when I run your example, I get the same result. In the Raku REPL:

```
> 'abc' ~~ m/(\w)+ {say $/} bc /;
｢abc｣
 0 => ｢a｣
 0 => ｢b｣
 0 => ｢c｣
｢ab｣
 0 => ｢a｣
 0 => ｢b｣
 0 => ｢c｣
｢a｣
 0 => ｢a｣
 0 => ｢b｣
 0 => ｢c｣
>
```

However, all the `$0` values appear to be intermediates, destined never to be assigned to a variable outside the scope of the smart-match. Here's what happens when I ask for `$/` if there's a match:

```
> say $/ if 'abc' ~~ m/(\w)+ {say $/} bc /;
｢abc｣
 0 => ｢a｣
 0 => ｢b｣
 0 => ｢c｣
｢ab｣
 0 => ｢a｣
 0 => ｢b｣
 0 => ｢c｣
｢a｣
 0 => ｢a｣
 0 => ｢b｣
 0 => ｢c｣
｢abc｣
 0 => ｢a｣
>
```

The only difference between the return from the second example when compared to the return from the first example are the last two lines:

 **｢abc｣
 0 => ｢a｣**

After performing the match above, asking Raku to `say` the `$/` variable produces the same result:

```
> say $/;
｢abc｣
 0 => ｢a｣
>
```

So it seems that Raku drops intermediate values at the conclusion of analyzing a regex, not before. An instructive SO post regarding this matter (from @jnthn ) may be found here:

https://stackoverflow.com/a/64407931/7270649

**Comment by juliodcs** [comment]
> However, all the $0 values appear to be intermediates, 
> destined never to be assigned to a variable outside the scope of the smart-match

While I can see that $0 could have intermediate values, I would expect them to be valid/consistent through the regex execution.

Consider this simpler case (without multiple $0 groups, just one $0 group):

    raku -e "say  'abc' ~~ m/(\w+) <?{$0.ends-with: 'b'}>/"
    ｢ab｣
     0 => ｢ab｣

This of course matches. 

However, If I try to do something _equivalent_ (same _contept_) with a repeated capturing group:

    raku -e "say  'abc' ~~ m/(\w)+ <?{$0.tail eq 'b'}>/"
    False

It can never match because $0 keeps all captured values.

I think that referencing capturing groups inside the regex is a common operation, so I was expecting (perhaps wrongly?) the $0 values to be updated.

**Comment by jubilatious1** [comment]
> However, If I try to do something equivalent (same contept) with a repeated capturing group { `code here` } It can never match because $0 keeps all captured values.

Unless I've misunderstood terribly, I can get the code you presented to work simply by replacing `tail` with `head`:

```
> say  'abc' ~~ m/(\w)+ <?{$0.head eq 'a'}>/
｢abc｣
 0 => ｢a｣
 0 => ｢b｣
 0 => ｢c｣
> say  'abc' ~~ m/(\w)+ <?{$0.head eq 'b'}>/
｢bc｣
 0 => ｢b｣
 0 => ｢c｣
> say  'abc' ~~ m/(\w)+ <?{$0.head eq 'c'}>/
｢c｣
 0 => ｢c｣
>
```

Remember for better or worse, the regex engine scans from left to right. So when the engine shifts right, `b` becomes the new `head`, not the new `tail`.

**Comment by juliodcs** [comment]
I believe that example is a little tricky. It seems to work but it does not.

If it "works", it is because the match fails and the pos is incremented +1, and that makes it match.

So it starts with pos = 0

    'abc' ~~ m/(\w)+ <?{$0.head eq 'a'}>/
    ^
     \____ pos

does *not* match with pos=0, so pos is incremented

So now we test the string with pos = 1

    'abc' ~~ m/(\w)+ <?{$0.head eq 'a'}>/
      ^
       \____ pos

Now it matches because 'bc' matches `(\w)+` and also $0.head equals 'b'.

But in the end that was a side effect of increasing pos. 

In fact, the capturing groups are added at the end of the array as seen on this example (I added an extra outter capturing group so that the inner array is displayed correctly)

```
> say 'abc' ~~ /((\w)+) {say $0} <?{False}>/
｢abc｣
 0 => ｢a｣
 0 => ｢b｣
 0 => ｢c｣
｢ab｣
 0 => ｢a｣
 0 => ｢b｣
｢a｣
 0 => ｢a｣
｢bc｣   # pos is incremented
 0 => ｢b｣
 0 => ｢c｣
｢b｣
 0 => ｢b｣
｢c｣    # pos is incremented
 0 => ｢c｣
Nil

```

**Comment by jubilatious1** [comment]
I think we need to back up a step and talk about the overall goal here. Lets take your last example, `say 'abc' ~~ /((\w)+) {say $0} <?{False}>/` . If I look at `$/` and `$0` afterwards, I get the following:

```
> say $/
Nil
> say $0
Nil
```
Now we both see the same intermediate results, but if at the end of the day `$/` and `$0` are both `Nil`, this means that those particular (interpolating) regex constructs probably won't be useful for assigning to a variable for further analysis.

At this point I think I have to contend that the regex engine is working properly, simply because it has maximal utility. My understanding is that when you or @raiph write a `(\w)` capture, you're asking for a single character capture--nothing more. Below:

```
> say  'abc' ~~ m/ \w /
｢a｣
> say  'abc' ~~ m:g/ \w /
(｢a｣ ｢b｣ ｢c｣)
> say  'abc' ~~ m/ \w+ /
｢abc｣
> say  'abc' ~~ m/ (\w)+ /
｢abc｣
 0 => ｢a｣
 0 => ｢b｣
 0 => ｢c｣
```

This makes sense to me. I access the last match/capture like so:

```
> say $/
｢abc｣
 0 => ｢a｣
 0 => ｢b｣
 0 => ｢c｣
> say $0
[｢a｣ ｢b｣ ｢c｣]
> say $0[0], $0[1], $0[2];
｢a｣｢b｣｢c｣
```

If I want to capture an exact number of separate, consecutive `\w` characters, I do the following:

```
> say  'abc' ~~ m/ (\w)**1 /
｢a｣
 0 => ｢a｣
> say  'abc' ~~ m/ (\w)**2 /
｢ab｣
 0 => ｢a｣
 0 => ｢b｣
> say  'abc' ~~ m/ (\w)**3 /
｢abc｣
 0 => ｢a｣
 0 => ｢b｣
 0 => ｢c｣
```

If I know I have exactly three `\w` characters and I want to capture them separately in separate `$0`, `$1`, `$2` variables, I do the following:

```
> say 'abc' ~~ m/ (\w) (\w) (\w) /
｢abc｣
 0 => ｢a｣
 1 => ｢b｣
 2 => ｢c｣
```

Note, the results above are all distinct from asking the regex engine to match/capture `(\w+)`, like so:
```
> say  'abc' ~~ m/ (\w+) /
｢abc｣
 0 => ｢abc｣
```
Finally, it seems to me that what you're really looking for is the `:exhaustive` adverb (abbreviated `:ex`), which, in conjunction with a `(\w+)` capture, will return matches/captures like so:

```
> say  'abc' ~~ m:ex/ \w+ /
(｢abc｣ ｢ab｣ ｢a｣ ｢bc｣ ｢b｣ ｢c｣)
> say $/.Str
abc ab a bc b c
>
> say  'abc' ~~ m:ex/ (\w+) /
(｢abc｣
 0 => ｢abc｣ ｢ab｣
 0 => ｢ab｣ ｢a｣
 0 => ｢a｣ ｢bc｣
 0 => ｢bc｣ ｢b｣
 0 => ｢b｣ ｢c｣
 0 => ｢c｣)
> say $/.Str
abc ab a bc b c
> say $0.Str
abc
```

Looking at the return from `:exhaustive`, you can get a better idea how regex backtracking works. First `abc` matches, then the engine backtracks to the `ab` match and again to the `a` match. The position is incremented and `bc` matches, then backtracking occurs to the `b` match. The position increments again and the `c` matches, at which point the engine stops because the end of the string has been reached.

[Now if there is something odd to this whole story, I would say in the last smart-match line `say  'abc' ~~ m:ex/(\w+)/;` (i.e. the line with capturing), that the two-at-a-time `$0` display seems a little unusual.].

Ref: https://docs.raku.org/language/regexes#Exhaustive

**Comment by juliodcs** [comment]
> I think we need to back up a step and talk about the overall goal here. Lets take your last example, `say 'abc' ~~ /((\w)+) {say $0} <?{False}>/` . If I look at `$/` and `$0` afterwards, I get the following:
> 
> ```
> > say $/
> Nil
> > say $0
> Nil
> ```
> 
> Now we both see the same intermediate results, but if at the end of the day `$/` and `$0` are both `Nil`, this means that those particular (interpolating) regex constructs probably won't be useful for assigning to a variable for further analysis.
> 
> 

Well that last example was just an example to show the intermediate capturing. Obviously that is not very useful by itself :-)

Anyways, considering that Raku has the possibility to use boolean assertions on the regexes, one could be interested in the intermediate values of $0, and not necessary on the final result.

> If I know I have exactly three \w characters and I want to capture them separately in separate $0, $1, $2 variables

Sounds logical, but what happens when you don't know the upper limit to match/capture?

Let's consider a different example, which will clarify this a little, I hope.

Even if this example could be done without a regex, please forget that for a moment :)

Consider a sequence of (numbers + one space). We try to find the list of numbers that sum up to 28.

The input is `"1 2 3 4 5 6 7 8 "` and the result should be `[｢1｣ ｢2｣ ｢3｣ ｢4｣ ｢5｣ ｢6｣ ｢7｣]`

So this would be the regex:

```
"1 2 3 4 5 6 7 8 " ~~ m/
    ^ [(\d) \s]+ <?{$0.sum == 28}>
    .* # done, no need to see more
/;

say $0;
# Nil
```

Of course, this fails because of the greedy quantifier and the way $0 behaves. 

The $0 gets all numbers, including the last one (8). The sum is not 28 so it backtracks a number, but $0 still has all the numbers, so the sum can never be 28, and it fails.

Consider now the following example which uses a workaround (adds an extra capturing group sorounding the quantizier)

```
"1 2 3 4 5 6 7 8 " ~~ m/
    ^ ([(\d) \s]+) <?{$0[0].sum == 28}> 
    .* # done, no need to see more
/;


say $0[0];
# [｢1｣ ｢2｣ ｢3｣ ｢4｣ ｢5｣ ｢6｣ ｢7｣]
```

This works. 

So to sum up:

- To my mind, the 'natural' case should behave the same as the one with the workaround.
- Even if you capture something, that doesn't necessary mean that you need it as a 'final' result. Since raku has boolean assertions, you may be interested in consistent intermediate results to use them with the assertions (or some other logic)

**Comment by jubilatious1** [comment]
This seems to work as expected:

```
> "1 2 3 4 5 6 7 8 " ~~ m/[(\d) \s]+ <?{ $0.sum == 28}> /;
False

> "1 2 3 4 5 6 7 8 " ~~ m/[(\d) \s]+? <?{ $0.sum == 28}> /;
｢1 2 3 4 5 6 7 ｣
 0 => ｢1｣
 0 => ｢2｣
 0 => ｢3｣
 0 => ｢4｣
 0 => ｢5｣
 0 => ｢6｣
 0 => ｢7｣
```
And this seems to work as expected, once you understand (from above) that "greediness" rules all:

```
> "1 2 3 4 5 6 7 8 " ~~ m/[(\d) \s]+ { say $0.sum } /;
36
> "1 2 3 4 5 6 7 8 " ~~ m/[(\d) \s]+ <?{ $0.sum == 36}> /;
｢1 2 3 4 5 6 7 8 ｣
 0 => ｢1｣
 0 => ｢2｣
 0 => ｢3｣
 0 => ｢4｣
 0 => ｢5｣
 0 => ｢6｣
 0 => ｢7｣
 0 => ｢8｣
```

In fact, the first time I see where the code immediately above doesn't work is when the regex engine switches to the next position rightward. No matter, we can pin the engine to the zero position:

```
> "1 2 3 4 5 6 7 8 " ~~ m/[(\d) \s]+ <!{ $0.sum == 36}> /;
｢2 3 4 5 6 7 8 ｣
 0 => ｢2｣
 0 => ｢3｣
 0 => ｢4｣
 0 => ｢5｣
 0 => ｢6｣
 0 => ｢7｣
 0 => ｢8｣
> "1 2 3 4 5 6 7 8 " ~~ m:pos(0)/[(\d) \s]+ <!{ $0.sum == 36}> /;
False
> "1 2 3 4 5 6 7 8 " ~~ m:pos(0)/[(\d) \s]+ <?{ $0.sum == 36}> /;
｢1 2 3 4 5 6 7 8 ｣
 0 => ｢1｣
 0 => ｢2｣
 0 => ｢3｣
 0 => ｢4｣
 0 => ｢5｣
 0 => ｢6｣
 0 => ｢7｣
 0 => ｢8｣
```

Once I understand pinning, I can go to town with the boolean assertion:

```
> "1 2 3 4 5 6 7 8 " ~~ m:pos(0)/[(\d) \s]+ <?{ $0.sum == 36}> \d\s /;
｢1 2 3 4 5 6 7 8 ｣
 0 => ｢1｣
 0 => ｢2｣
 0 => ｢3｣
 0 => ｢4｣
 0 => ｢5｣
 0 => ｢6｣
 0 => ｢7｣
> say $0
[｢1｣ ｢2｣ ｢3｣ ｢4｣ ｢5｣ ｢6｣ ｢7｣]
```

Aha! I get the `$0` I desire! Of course, the above **sum** for `$0` seems wrong, but as noted earlier it's a consequence of the greedy capture (and may in fact be unavoidable). Let's fix that with a frugal (non-greedy) quantifier:

```
> "1 2 3 4 5 6 7 8 " ~~ m:pos(0)/[(\d) \s]+? <?{ $0.sum == 36}> \d\s /;
False
> "1 2 3 4 5 6 7 8 " ~~ m:pos(0)/[(\d) \s]+? <?{ $0.sum == 28}> \d\s /;
｢1 2 3 4 5 6 7 8 ｣
 0 => ｢1｣
 0 => ｢2｣
 0 => ｢3｣
 0 => ｢4｣
 0 => ｢5｣
 0 => ｢6｣
 0 => ｢7｣
> say $0
[｢1｣ ｢2｣ ｢3｣ ｢4｣ ｢5｣ ｢6｣ ｢7｣]
> "1 2 3 4 5 6 7 8 " ~~ m:pos(0)/[(\d) \s]+? <?{ $0.sum == 28}> /;
｢1 2 3 4 5 6 7 ｣
 0 => ｢1｣
 0 => ｢2｣
 0 => ｢3｣
 0 => ｢4｣
 0 => ｢5｣
 0 => ｢6｣
 0 => ｢7｣
> "1 2 3 4 5 6 7 8 " ~~ m/[(\d) \s]+? <?{ $0.sum == 28}> /;
｢1 2 3 4 5 6 7 ｣
 0 => ｢1｣
 0 => ｢2｣
 0 => ｢3｣
 0 => ｢4｣
 0 => ｢5｣
 0 => ｢6｣
 0 => ｢7｣
```
Okay, `$0` looks like what we want. Actually we got the correct `$0` earlier but it makes more sense in this form. The take-home lesson for me is greediness overrules boolean assertions, that is to say, you can't rely on backtracking to alter capture lengths (or sums, etc.) if you use a greedy as opposed to frugal (non-greedy) quantifier.

**Comment by raiph** [comment]
Please know you don't have to reply to this comment either; I'm hoping this is my last in this issue -- at least unless/until someone who has looked at the code in the regex engine and/or a core dev who knows the engine well enough to comment, tells us what's happening.

But I wanted to address a couple things.

> At this point I think I have to contend that the regex engine is working properly, simply because it has maximal utility.

You don't *have* to, but fair enough. :)

I think the principle of prioritizing maximal utility is right for Raku, all other things being more or less equal.

And I could see that applying in the sense of Rakudo implementation and our limited resources. Presuming my analysis (in the last bit of my opening comment) of what's happening during backtracking is correct, it might be that it's so expensive to compute which trailing elements of a `List` of captures in a capture need to be `:delete`d, and/or to `:delete` them, that it's best left undone. And I mean "expensive" in terms of either dev time or run time performance.

Another interpretation of "maximal utility" would be that there's something that can now be done that would no longer be possible if capture `List`s were to be properly truncated during backtracking. I don't think you're claiming that, but if you were, and you're right, then that would of course be pertinent too.

It currently seems to me that:

* Open ended quantification (`+`, `*`, and `**` with either a range or dynamic variability) of a capture; and

* Failing or succeeding parse progress based on some computation of what such captures have captured; and

* Greedy backtracking;

are fundamental both individually and in combination, and I'm not currently thinking there *is* another clean way to do anything they can reasonably and cleanly be used for.

That all said, maybe I'm wrong, in which case great. :)

Or maybe this is just an unfortunate Rakudo bug that we need to live with for now, and perhaps forever, in which case, OK.
